### ISSUE TITLE
Attractor bug on Qwen3.6:27b MQ4 with unusual prompt

### ISSUE THREAD

**Issue by darkamgine** [Kaden-Schutt/hipfire#213]
Hosting this model and a friend managed to trigger an attractor bug with an unusual prompt: `what are the must-try foods when in malaysia?`

I tried several variations, swapping out "malaysia" for other countries and didn't get any issues.

I tried on a couple of custom settings tweaking some of the variables like temperature, presence penalty, and repetition penalty but still seeing repetitive blocks on both the thinking and the main responses.

I tried running `Qwen3.6-27B-UD-Q4_K_XL` on llama.cpp with the same settings and didn't see this happening.

Running on Windows on commit `e685cf` (~2 days ago) and compiled executables with cargo manually. No dflash or any extra environment variables.

I know there were talks about attractor bugs on moe models, so hopefully this report is unique as I didn't see anything about it affecting dense models before.

Settings and screenshots provided for reference, happy to provide more detail as needed.

As an aside, decode on hipfire was 34t/s compared to 6.8t/s on llama with this prompt on an 9070xt, which is amazing when it works properly. I noticed the decode speed is faster than hipfire from a week ago by ~30+%, though it seems to use more cpu (~10%) and regular ram (~7gb) than before? (not sure if im misremembering)

Setting used to generate below results:
```
  "qwen3.6:27b": {
    "thinking": "on",
    "preserve_thinking": true,
    "max_tokens": 4096,
    "max_think_tokens": 1500,
    "max_seq": 65536,
    "temperature": 0.5,
    "top_p": 0.95,
    "top_k": 20,
    "min_p": 0.0,
    "presence_penalty": 0.0,
    "repetition_penalty": 1.0
  }
```
Repetition in thinking block, cut off after 1500 tokens
<img width="1337" height="833" alt="Image" src="https://github.com/user-attachments/assets/556f058f-3771-4368-804a-cee2c4421c31" />
Repetition in main response, cut off after 4096 tokens
<img width="1345" height="947" alt="Image" src="https://github.com/user-attachments/assets/89d52f70-5fc6-4599-9115-731f4b161ee1" />
No issues on llama.cpp with Unsloth quants of 27B
<img width="845" height="787" alt="Image" src="https://github.com/user-attachments/assets/624a8e04-8685-4b96-b23f-69d10cd7ae8f" />

**Comment by darkamgine** [comment]
When running on mostly default settings without max_think_tokens, it gets stuck at the thinking loop and never finishes:

```
"qwen3.6:27b": {
  "max_tokens": 4096,
  "max_seq": 65536
}
```

<img width="781" height="671" alt="Image" src="https://github.com/user-attachments/assets/b53b1bd5-c5ae-4061-bd1a-dd7fd0eb1d99" />

**Comment by darkamgine** [comment]
one more weird result, when switching the country to `thailand`, it gives up after 322 tokens during the thinking phase, but the output looks coherent until then.

<img width="907" height="799" alt="Image" src="https://github.com/user-attachments/assets/b556e7e1-160e-42b3-8d40-a48881fa739b" />

**Comment by unverbraucht** [comment]
@darkamgine could you retry with latest master? We fixed one correctness bug, but more importantly we turned off repeat penalty by default. This fixes a lot of reasoning looping issues, or probably all of them :)

**Comment by 0x00cl** [comment]
> one more weird result, when switching the country to `thailand`, it gives up after 322 tokens during the thinking phase, but the output looks coherent until then.

Regarding this, another user on Discord stumbled into this same problem with another prompt `hipfire run qwen3.6:27b "Qu'est-ce que la physique quantique?"` It would output nothing and it would only generate 256 thinking tokens and stop.

After a lot of testing, I don't know if its a bug or not, it simply has to do with the default parameters that hipfire provides for the model. If you run the query with the suggested parameters by Qwen team, it will work: `hipfire run qwen3.6:27b "Qu'est-ce que la physique quantique?" --max-tokens 2048 --temperature 1.0 --top-p 0.95 --top-k 20 --min-p 0.0 --repeat-penalty 1.0`

**Comment by darkamgine** [comment]
@unverbraucht I can confirm after building the latest from master that the response for the original prompt for `malaysia` is able to complete normally now!

However, the prompt for `thailand` (`what are the must-try foods when in thailand?`) is still resulting in a cut-off response during the thinking phase between 200-300 tokens. This is happening with hipfire's default settings, with my settings above, and with Qwen's recommended settings per @0x00cl .

**Comment by 0x00cl** [comment]
I tried your prompt like the following `hipfire run qwen3.6:27b "what are the must-try foods when in thailand?" --max-tokens 2048 --temperature 1.0 --top-p 0.95 --top-k 20 --min-p 0.0 --repeat-penalty 1.0` and works fine, though I did try to test it using open-webui and although I adjusted the parameters of the model, the response gets all messed up and answers something different to the `hipfire run` command, so open-webui may be sending something else or idk why I can't repeat it.

Anyways, I haven't seen this kind of behaviour with llama.cpp and Unsloth quants either.

For what is worth I asked GLM 5.1 to analyze possible bugs in the code regarding this and this was the answer (I wish I could understand how llms worked to give my input and if the answer makes sense):

---
# Analysis: Premature Generation Stop During Thinking

I identified three bugs and one design concern that together explain the behavior you're seeing.

---
# Bug 1 (Primary): Hardcoded top-K=20 traps the model inside the think block

Location: `kernels/src/sample_top_p.hip:36`, `crates/hipfire-runtime/src/llama.rs:4509`

The sampler uses a hardcoded TOP_K = 20 with no configuration knob. The selection happens on raw logits before temperature scaling. Here's why this is the core problem:
1. While reasoning, the model's "continue thinking" logits dominate the top-20 positions
2. The </think closing token sits outside the top-20 in raw logit space
3. Top-K=20 permanently excludes it before softmax even runs
4. At the effective temperature of ~0.246 (0.3 * 0.82 TEMP_CORRECTION at `cli/index.ts:19`), softmax over the remaining 20 tokens is extremely peaked (~90%+ mass on the top-1)
5. The model cannot escape the think block through sampling alone

This is self-reinforcing: reasoning continues → attention reinforces reasoning logits → </think stays outside top-20 → model keeps reasoning. With repeat_penalty=1.0 (daemon default at `daemon.rs:947`), there's zero suppression to help.

Why llama.cpp doesn't do this: llama.cpp uses top-K=40 by default and applies temperature before top-K selection, so the </think token's probability is amplified by 1/temp scaling before the cutoff. hipfire selects top-K from raw (unscaled) logits, which means temperature has no effect on which tokens enter the candidate pool.

---
# Bug 2: EOS/EOT stop check fires before think-cap logic

Location: `daemon.rs:3910-3912` (AR path), `daemon.rs:2726` (DFlash path)

The decode loop ordering is:
1. Forward scratch (KV write)
2. EOS check → break              ← fires FIRST
3. is_terminator() check → break  ← fires FIRST  
4. max_think_tokens enforcement   ← too late
5. Sample next token

If the model emits <|im_end|> (EOS) or `` (EOT) during thinking — which is plausible when top-K=20 prevents the model from closing </think naturally and forces it into a corner of its distribution — generation stops immediately at step 2/3, completely bypassing the think-cap force-close logic at step 4. The result: an unclosed think block.

The is_terminator() function (`tokenizer.rs:559-561`) catches both <|im_end|> and . For Qwen 3.6,  is registered as eot_id. If the model emits this token while reasoning (because the actual </think token was excluded by top-K=20), the engine interprets it as "generation complete" and stops.

---
# Bug 3: Think-cap force-close can fail silently when budget is exhausted

Location: `daemon.rs:3938-3973`

When max_think_tokens is hit, the force-close logic computes:

let budget_left = max_tokens.saturating_sub(generated);
let take = close_tokens.len().min(budget_left);

If generated >= max_tokens (e.g., all tokens consumed by thinking), then budget_left = 0, take = 0, no close tokens are emitted, and the loop exits via the if generated >= max_tokens { break; } check with an unclosed thought. This happens when:
- The client sends reasoning.effort: "low" (maps to max_think_tokens=256 per `cli/index.ts:1820`)
- AND max_tokens is small (e.g., 256-300)
- 
---
# Design Concern: TEMP_CORRECTION = 0.82

Location: `cli/index.ts:19`

All temperatures are silently multiplied by 0.82 before sending to the daemon. So "temperature 0.3" in the UI becomes 0.246 at the engine level. This makes sampling even more greedy, compounding Bug 1 by further concentrating probability mass on the top-1 token from the already-narrow top-20 pool.

---
# Proposed Fix Plan

1. Make top-K configurable — add top_k to SamplerConfig and the GPU kernel, default to 40 (matching llama.cpp). This is the single highest-impact fix.
2. Apply temperature before top-K selection — currently top-K runs on raw logits. Temperature-scaling logits before top-K would allow 1/T amplification to promote </think into the candidate set even when its raw logit is outside top-20. Alternatively, increase top-K to a value large enough that this is rarely an issue (e.g., 200).
3. Reorder EOS checks relative to think-cap — move the think-cap check before the EOS/EOT stop conditions, so that if the model emits a terminator inside a think block, the engine has a chance to force-close the thought first rather than stopping with an unclosed block.
4. Reserve budget for the force-close sequence — the think-cap force-close should guarantee enough tokens from max_tokens to emit </think\n, even if it means reducing the effective think budget by the close sequence length (typically 2-4 tokens).
5. Remove or document TEMP_CORRECTION — the silent 0.82 multiplier is surprising and reduces the effective temperature range.

**Comment by Kaden-Schutt** [comment]
Partial resolution confirmed: the **"malaysia" loop** is fixed by the `repeat_penalty` 1.3→1.0 default (PR #267, `9b4ab74a`) — which you confirmed in the 2026-05-21 comment. The remaining symptom (output **cut off mid-`<think>`** at ~200–300 tokens) is a separate bug: a terminator (`<|im_end|>`/eot) sampled *inside* the open `<think>` block breaks the decode loop before the think-cap force-close fires. Recent bounded-thinking / premature-stop work has landed since (`0ec71c30`, `a7dcfb0d`) that may already address it — could you re-test the "thailand" prompt on latest master? If it still truncates, the remaining fix is to route a terminator-inside-think through the existing `</think>` force-close path. Keeping open pending your re-test (needs the Qwen3.6-27B MQ4 model to reproduce the exact symptom).

_Full feasibility analysis + step-by-step plan in the triage report: `docs/issue-triage-2026-06-03.md` (PR #394). Triaged 2026-06-03 against `origin/master`._
